lms/djangoapps/courseware/courses.py

```diff
@@ -405,17 +410,50 @@ def get_course_date_blocks(course, user):
     if certs_api.get_active_web_certificate(course):
         block_classes.insert(0, CertificateAvailableDate)
 
-    blocks = (cls(course, user) for cls in block_classes)
+    blocks = [cls(course, user) for cls in block_classes]
+    blocks.extend(get_course_assignment_due_dates(
+        course, user, request, num_return=num_assignments, include_past_dates=include_past_dates))
 
-    def block_key_fn(block):
-        """
-        If the block's date is None, return the maximum datetime in order
-        to force it to the end of the list of displayed blocks.
-        """
-        if block.date is None:
-            return datetime.max.replace(tzinfo=pytz.UTC)
-        return block.date
-    return sorted((b for b in blocks if b.is_enabled), key=block_key_fn)
+    return sorted((b for b in blocks if b.date and (b.is_enabled or include_past_dates)), key=date_block_key_fn)
+
+
+def date_block_key_fn(block):
+    """
+    If the block's date is None, return the maximum datetime in order
+    to force it to the end of the list of displayed blocks.
+    """
+    return block.date or datetime.max.replace(tzinfo=pytz.UTC)
+
+
+def get_course_assignment_due_dates(course, user, request, num_return=None, include_past_dates=False):
+    """
+    Returns a list of assignment (at the subsection/sequential level) due date
+    blocks for the given course. Will return num_return results or all results
+    if num_return is None in date increasing order.
+    """
+    store = modulestore()
+    all_course_dates = get_dates_for_course(course.id, user)
+    date_blocks = []
+    for (block_key, date_type), date in all_course_dates.items():
+        if date_type == 'due':
+            item = store.get_item(block_key)
```

Comment:
Does it return a lot of things that aren't `sequential` blocks? If so, it might be faster to derive the category from the block_key itself (`block_key.block_type`), and only make the modulestore `get_item` call after you know that it's a sequential (modulestore queries are often expensive).

Reply:
It looks like about 2/3 of what we have in edx_when right now are sequentials, but if that is a better method, I'm happy to switch it. I honestly just didn't know the block_key had a block_type property. Thanks Dave!